### PR TITLE
feature: S3C-4073 Add new probe server

### PR DESCRIPTION
**Pull request by miniscruff**
* JsDocs for arsenal error
* ProbeServer as a replacement to HealthProbeServer

### CODE REVIEW

File: tests/unit/network/probe/ProbeServer.spec.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    }).end();
+}
+
+describe('network.probe.ProbeServer', () => {
+    /** @type {ProbeServer} */
+    let server;
+
+    beforeEach(done => {
+        server = new ProbeServer({ port: 4042 });
+        server._cbOnListening = done;
```
Comment (naren-rajendran) [inline tests/unit/network/probe/ProbeServer.spec.js:29]: what does this do?

File: tests/unit/errors.js [hunk 2]
```diff
@@ -26,4 +26,17 @@ describe('Errors: ', () => {
...
+        const fakeRes = {
+            writeHead: code => assert.strictEqual(code, 404),
+            end: msg => {
+                assert.strictEqual(
+                    msg,
+                    errors.NoSuchEntity.toString(),
+                );
+            },
+        };
+        errors.NoSuchEntity.writeResponse(fakeRes);
```
Comment (naren-rajendran) [inline tests/unit/errors.js:40]: Neat fake response 👍🏽

File: tests/unit/network/probe/ProbeServer.spec.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    it('error on bad method', done => {
+        makeRequest('POST', '/unused', (err, res) => {
+            assert.ifError(err);
+            assert.strictEqual(res.statusCode, 405);
+            done();
+        });
+    });
+
+    it('error on missing request', done => {
```
Comment (naren-rajendran) [inline tests/unit/network/probe/ProbeServer.spec.js]: missing request?

File: tests/unit/network/probe/ProbeServer.spec.js (hunk 1 above)
Comment (miniscruff) [reply tests/unit/network/probe/ProbeServer.spec.js:29]: I think this is a callback used for when the server is up and listening, I believe

File: tests/unit/network/probe/ProbeServer.spec.js (hunk 3 above)
Comment (miniscruff) [reply tests/unit/network/probe/ProbeServer.spec.js]: Like it is missing the /unused request endpoint.
Edit: Could maybe use a better name...

File: lib/network/http/server.js [hunk 4]
```diff
@@ -203,8 +204,8 @@ class Server {
     /**
      * Function called when request received
      *
-     * @param {http.IncomingMessage|https.IncomingMessage} req - Request object
-     * @param {http.ServerResponse|https.ServerResponse} res - Response object
+     * @param {http.IncomingMessage} req - Request object
+     * @param {http.ServerResponse} res - Response object
```
Comment (anurag4DSB) [inline lib/network/http/server.js:208]: Is it just standard practice to write jsdoc or we do not support receiving HTTPs requests?

Comment (anurag4DSB) [reply lib/network/http/server.js:208]: Same at other places.

File: lib/network/probe/ProbeServer.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+const DEFAULT_LIVE_ROUTE = '/_/live';
+const DEFAULT_READY_ROUTE = '/_/live';
+const DEFAULT_METRICS_ROUTE = '/_/metrics';
+
+/**
+ * ProbeDelegate is used to determine if a probe is successful or
+ * if any errors are present.
+ * If everything is working as intended, it is a no-op.
+ * Otherwise, return a string representing what is failing.
+ * @callback ProbeDelegate
```
Comment (anurag4DSB) [inline lib/network/probe/ProbeServer.js:14]: How is ProbeDelegate a callback?

**Review by anurag4DSB — APPROVED** [review]
Just curious about some things.
But otherwise looks good to me.

File: lib/network/probe/ProbeServer.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    _onRequest(req, res) {
+        const log = this.logging.newRequestLogger();
+        log.debug('request received', { method: req.method, url: req.url });
+
+        if (req.method !== 'GET') {
+            errors.MethodNotAllowed.writeResponse(res);
+            return;
+        }
+
+        if (!(req.url in this._handlers)) {
```
Comment (tmacro) [inline lib/network/probe/ProbeServer.js]: Functionally the same, but `if (!this._handlers.includes(req.url))` avoids the nested `()`

File: tests/unit/network/probe/ProbeServer.spec.js (hunk 3 above)
Comment (naren-rajendran) [reply tests/unit/network/probe/ProbeServer.spec.js]: invalid route?

File: lib/network/http/server.js (hunk 4 above)
Comment (miniscruff) [reply lib/network/http/server.js:208]: There is no https request, I looked it up, https server uses the same response and request as http.

File: lib/network/probe/ProbeServer.js (hunk 5 above)
Comment (miniscruff) [reply lib/network/probe/ProbeServer.js:14]: There is not jsdoc for delegates so I used callback here as the closest but also included the @returns to specify you can return values like a normal delegate.

File: lib/network/probe/ProbeServer.js (hunk 6 above)
Comment (miniscruff) [reply lib/network/probe/ProbeServer.js]: Objects do not have an includes method, I can do:
```suggestion
        if (!this._handlers.hasOwnProperty(req.url)) {
```

File: lib/network/probe/ProbeServer.js (hunk 5 above)
Comment (miniscruff) [reply lib/network/probe/ProbeServer.js:14]: https://jsdoc.app/tags-callback.html says you can specify the return value, seems JsDoc does not differentiate between a callback ( no return value ) and a delegate ( one or more values ) and just has one group. But we can at least use naming to separate them.

File: lib/network/probe/ProbeServer.js (hunk 6 above)
Comment (miniscruff) [reply lib/network/probe/ProbeServer.js]: I could use a Map instead and then use the `has` method. Would prevent any weird stuff from the built-in object methods and be more aligned with it being a map and not an Object with methods and things.

**Review by naren-rajendran — APPROVED** [review]

File: lib/network/probe/ProbeServer.js (hunk 6 above)
Comment (miniscruff) [reply lib/network/probe/ProbeServer.js]: @tmacro switched to a Map, let me know your thoughts

File: lib/network/probe/ProbeServer.js (hunk 5 above)
Comment (anurag4DSB) [reply lib/network/probe/ProbeServer.js:14]: This reminds me of our discussion of specifying `@returns {undefined}` or `@returns {callback}`.
I think this info completes that loop :)

File: lib/network/probe/ProbeServer.js (hunk 6 above)
Comment (tmacro) [reply lib/network/probe/ProbeServer.js]: TIL about Map :+1:

Comment (miniscruff) [reply lib/network/probe/ProbeServer.js]: ooh Maps are super great, would recommend.

**Pull request merged**